site/docs/types/type_variations.md

```diff
@@ -0,0 +1,11 @@
+# Type Variations
+
+Since Substrait is designed to work in both logical and physical contexts, there is a need to support extended attributes in the physical context. Different consumers may have multiple ways to present the same logical type. For example, an engine might support dictionary encoding a string or using either a row-wise or columnar representation of a struct. As such, there is the facility for specification users to express additional type variations for each logical type. These variations are expected to have the same logical properties as the canonical variation and are defined for each organization. The key properties of these variations are:
+
+| Property          | Description                                                  |
+| ----------------- | ------------------------------------------------------------ |
+| Parent Type       | The parent type this variation belongs to. Variations can only be expressed for simple types and wild-carded compound types (e.g. i8 or varchar(*)). |
+| Name              | The name used to reference this type. Should be unique within type variations for this parent type within an organization. |
+| Description       | A human description of the purpose of this type variation    |
+| Function Behavior | **Inherits** or **Independent**: Whether this variation supports functions using the canonical variation or whether functions should be resolved independently. For example if one has the function `add(i8,i8)` defined and then defines an i8 variation, can the i8 variation field be bound to the base `add` operation (inherits) or does a specialized version of `add` need to be defined specifically for this type variation (independent). Defaults to inherits. |
```

Comment:
If "function behavior" is `independent` then I would argue it is not a variation but a separate logical type at that point.  I may be misunderstanding the definition here though.  My thinking of "inherits" means "these two types have the same semantic meaning".  So "independent" would mean that the type has a different semantic meaning.

In other words.  I see two possible interpretations (neither of which are very appealing)

* Let there be a function F.  The term `inherits` means that if `x AS VARIATION v1 == x AS VARIATION v2` and `F(x) => y` then `y AS VARIATION v1 == y AS VARIATION v2`.
    * I don't like this interpretation because I think this should be true for every variation no matter what.
* If a consumer supports function F on type `X as variation v1` then it MUST support function F on type `X as variation v2`
    * I'm not a fan of this one because I think there will always be function signatures that aren't supported by one particular engine or another and the fact is unrelated to variations.  Plus, a query planner could always discover this via function introspection and presumably cast to a supported variation (or the consumer itself could support this).

Reply:
Imagine a situation where you have a physical implementation function that only works for dictionary encoded strings (but does not work for the "system default" strings). Both values are logically strings but only a subset of operations may be available for one physical representation. That is the purpose of independent. If you're using inherits, it means that a function defined for the base type would be available for the variation. 

To me, part of having physical variations is that there will likely be some operations that only work against one of the variations.

Reply:
Right.  I agree those situations may exist.  I just don't think Substrait should concern itself with defining which situations are valid or invalid.  Here are two concrete examples:

 * Arithmetic functions in Arrow work on regular numbers as well as dictionary encoded numbers.  This is an example where we support functions for all physical representations.
 * The system default timestamp representation may be a 64 bit integer.  An alternative is a packed struct with year/month/day/...  Field extraction (e.g. get_hour) on the former is expensive.  Field extraction on the latter is trivial.  Users typically want multiple fields (e.g. it's rare to just ask for the hour, often you want hour, minutes, and seconds).  So, an engine may choose to only implement the field extraction kernel for the packed struct representation, forcing a cast.

Is it an error, in Substrait's eyes, if an engine fails to support arithmetic functions on dictionary numbers?  Is it an error if an engine chooses to allow field extraction on the system default representation?

I don't think Substrait should care.